Be date spec comparator

treble37 opened the pull request:
For https://github.com/antonmi/espec/issues/111

- [x] Please namespace all the modules connected with this feature to separate module.
Let's say ESpec.DatesTimes. And place the code into dates_times directory.
- [x] Please add specs for different cases. Place them to 'spec/dates_times' folder.
- [x] Address Zeeker's recommendations
- [x] Please add README section which describes the cases
- [x] Return {result, delta} (https://github.com/antonmi/espec/pull/226#discussion_r143116826) ...?
- [x] Also, could we think about improvements of be(:>, ...) and be(:<, ...) mathers? So they use the same comparison logic.
- [x] Misc: Add extra error tests in dates_times/be_close_to.exs
- [x] use the Comparator.diff when dates are passed to the be assertion 
 (https://github.com/antonmi/espec/pull/226#issuecomment-341986229)
- [x] Add case for single keyword list (see https://github.com/antonmi/espec/pull/226#issuecomment-342240171 and https://github.com/antonmi/espec/pull/226#issuecomment-342240746)
- [x] clean up spec docs
- [x] Add equivalent ExUnit assertions...
- [x] update README with the new keyword list args in "be" assertion

antonmi commented:
Hi, @treble37 !
Sorry, I may have missed smth, but what is the status of this PR?
Is it ready for merging or you want to add smth?
Thanks!

treble37 commented:
Hi @antonmi,  just a few things...

1) I was hoping for feedback on the handling of the timezones wrt DateTime tests ... essentially what I did was backport a fair amount of the Elixir 1.5 functionality to make it compatible down to 1.3
2) Shall I also add the corresponding ExUnit tests as well in this PR?
3) Based on the feedback, I might make some changes then, depending....

antonmi commented:
Oh, sorry!
I will review soon

antonmi commented:
Hi, @treble37 !
Sorry again for such late review!
There are a few comments:
1. Please namespace all the modules connected with this feature to separate module.
Let's say ESpec.DatesTimes. And place the code into dates_times directory. 
So the structure will be:
lib/
  dates_times/
    comparator.ex (call the module ESpec.DatesTimes.Comparator, no need to separate Diff module)
	date_time_protocol.ex (So module will be ESpec.DatesTimes.DateTimeProtocol ) 
	types.ex (ESpec.DateTimes.Types )
	impl/
  	  date.ex
	  datetime.ex
	  naive_datetime.ex
	  time.ex
	extensions/
	  calendar_extension.ex
	  datetime_extension.ex
	  integer_extension.ex  
	
2. Please add specs for different cases. Place them to 'spec/dates_times' folder.
3. Please add README section which describes the cases

Also, could we think about improvements of be(:>, ...) and be(:<, ...) mathers? So they use the same comparison logic.

Thank you!

Zeeker commented:
@antonmi, @treble37 are you okay with me chiming in on the changes and also providing a review? I don't want to force myself into this. :)

antonmi commented:
@Zeeker you are very welcome!

treble37 commented:
Hi @Zeeker - your input would be welcome, I have a feeling there are things that can be improved, so please review away and don't be shy!

Zeeker commented:
I will, lot's of meetings today! 😄

antonmi commented:
@Zeeker , thank you very much for detailed comments!

treble37 commented on lib/espec/comparator/diff.ex:
```diff
@@ -0,0 +1,110 @@
+    end
+  end
+
+  defp do_diff(a, a, type),      do: zero(type)
```
Adjusted in accordance with https://github.com/antonmi/espec/pull/226#discussion_r143118644 (`If you use the do: syntax with functions and the line that makes up the function body is long, put the do: on a new line indented one level more than the previous line. `)

treble37 commented on lib/espec/comparator/diff.ex:
```diff
@@ -0,0 +1,110 @@
+  end
+  defp do_diff(a, b, :years) do
+    diff_years(a, b)
+  end
```
Adjusted in accordance with https://github.com/antonmi/espec/pull/226#discussion_r143118644 (`If you use the do: syntax with functions and the line that makes up the function body is long, put the do: on a new line indented one level more than the previous line. `)

treble37 commented on lib/espec/comparator/diff.ex:
```diff
@@ -0,0 +1,110 @@
+    diff_years(a, b)
+  end
+  defp do_diff(_, _, granularity) when not granularity in @units,
+    do: {:error, {:invalid_granularity, granularity}}
```
Adjusted in accordance with https://github.com/antonmi/espec/pull/226#discussion_r143118644 (`If you use the do: syntax with functions and the line that makes up the function body is long, put the do: on a new line indented one level more than the previous line. `)

treble37 commented on lib/espec/assertions/be_close_to.ex:
```diff
@@ -5,6 +5,12 @@ defmodule ESpec.Assertions.BeCloseTo do
   it do: expect(2).to be_close_to(1, 3)
   """
   use ESpec.Assertions.Interface
+  alias ESpec.Comparator.Diff
```
Space added between use and alias

treble37 commented on lib/espec/assertions/be_close_to.ex:
```diff
@@ -5,6 +5,12 @@ defmodule ESpec.Assertions.BeCloseTo do
+  alias ESpec.Comparator.Diff
+
+  defp match(subject, [value, {granularity, delta}] = data) do
+    result = abs(Diff.diff(subject, value, granularity)) <= delta
```
thanks, I like this and will use it.

treble37 commented on lib/espec/comparator/diff.ex:
```diff
@@ -0,0 +1,110 @@
+  defp do_diff(_, _, granularity) when not granularity in @units,
+    do: {:error, {:invalid_granularity, granularity}}
+
+  defp diff_years(a, a), do: 0
```
great catch Zeeker, will change

treble37 commented on lib/espec/comparator/diff.ex:
```diff
@@ -0,0 +1,110 @@
+    if a > b do
+      diff_years(end_date, start_date, 0)
+    else
+      diff_years(start_date, end_date, 0) * -1
```
Will change, thanks!

treble37 commented on lib/espec/comparator/diff.ex:
```diff
@@ -0,0 +1,110 @@
+          12
+      end
+    (year_diff*12)+month_diff
+  end
```
@Zeeker - the current implementation of diff_years that I think you're talking about (listed below) assumes a and b are date(times) converted to gregorian microseconds.....

```
  defp diff_years(a, b) do
    {start_date, _} = :calendar.gregorian_seconds_to_datetime(div(a, 1_000*1_000))
    {end_date, _} = :calendar.gregorian_seconds_to_datetime(div(b, 1_000*1_000))
    if a > b do
      do_diff_years(end_date, start_date, 0)
    else
      do_diff_years(start_date, end_date, 0) * -1
    end
  end
```

treble37 commented on lib/espec/assertions/be_close_to.ex:
```diff
@@ -5,6 +5,12 @@ defmodule ESpec.Assertions.BeCloseTo do
+
+  defp match(subject, [value, {granularity, delta}] = data) do
+    result = abs(Diff.diff(subject, value, granularity)) <= delta
+    {result, result}
```
Preliminary interface for providing a real computed delta with an error message added...

treble37 commented:
Ok @antonmi (and @Zeeker), this PR is getting close to done (other than some minor documentation cleanup), but I would like some feedback regarding your comment:

> Also, could we think about improvements of be(:>, ...) and be(:<, ...) mathers? So they use the same comparison logic.

Commit https://github.com/antonmi/espec/pull/226/commits/9b5d91d6bf989e4405d384acfa420e1b16e681a4 is my attempt to do this.  What do you think of this API? If you're fine with it, then I will add the necessary tests to implement.

antonmi commented:
Hi, @treble37 !

Thanks for your awesome work!
About `be` assertion:

What I meant was use the Comparator.diff when dates are passed to the `be` assertion:

Like this:
```elixir
  defp match(%Date{} = subject, [op, %Date{} = val]), do: match_date_times(subject, [op, val])
  defp match(%Time{} = subject, [op, %Time{} = val]), do: match_date_times(subject, [op, val])
  defp match(%DateTime{} = subject, [op, %DateTime{} = val]), do: match_date_times(subject, [op, val])
  defp match(%NaiveDateTime{} = subject, [op, %NaiveDateTime{} = val]), do: match_date_times(subject, [op, val])

  defp match_date_times(subject, [op, val]) do
    delta = subject |> Comparator.diff(val, :seconds)

    result = apply(Kernel, op, [delta, 0])
    {result, result}
  end
```

But your API implementation is also very cool!
So, my suggestion: Let's keep both!

Zeeker commented:
Hey @treble37,

I can only join antomni in saying thank you for your work and patience, and I'm glad I could be of some help.

I just have one final suggestion. In addition to the current `be` Implementation, you could add a case for a single keyword list, similar to how `have` can now receive values.

I'll link in a second, I'm on mobile right now.

Zeeker commented:
See commit ae7a4ca7c6daffd86ccc227a06498dbe72c935ed for details of what I mean

treble37 commented:
ok @Zeeker, regarding 
 your comment https://github.com/antonmi/espec/pull/226#issuecomment-342240171 this commit (https://github.com/antonmi/espec/pull/226/commits/ec3986d5355e70b5b05d02ebb52a5c38c2513e77) enables it

treble37 commented:
ok @antonmi and @Zeeker, this PR is now updated with the functionality and I addressed yours and Zeeker's comments (which I put in the checklist up top https://github.com/antonmi/espec/pull/226#issue-252002626), so I guess it's ready to merge unless you have any other comments?

I can bump the version to 1.4.7 and edit the CHANGELOG if you're ok with everything.

Zeeker commented:
Awesome work @treble37! Semantic versioning would suggest to bump the version to 1.5.0 since you introduce a backwards compatible feature. Those are just my two cents though. 🙂

antonmi commented:
Absolutely awesome! 
Thank you @treble37 !
I will review and merge soon!
Thank you @Zeeker for comments!

treble37 commented:
Of course you are right @Zeeker - version bumped to 1.5.0 (maybe I shouldn't have had that last 🍷 ;) )

antonmi commented:
ESpec 1.5.0 is live!
Thanks again to everyone!
